File: python/bf.py
import sys
import os


def change(tape, pointer, amount):
    while len(tape) <= pointer:
        tape.append(0)
    
    tape[pointer] = (tape[pointer] + amount) % 255


def increment(tape, pointer):
    change(tape, pointer, 1)


def decrement(tape, pointer):
    change(tape, pointer, -1)

def is_zero(tape, pointer):
    change(tape, pointer, 0)
    return tape[pointer] == 0


def main():
    if len(sys.argv) != 2:
        print('A simple BF intepreter.')
        print('Usage: \t ./bf.py program.bf')
        exit(0)

    filename = sys.argv[1]
    if not os.path.isfile(filename):
        print(f'Could not find file {filename}')
        exit(0)
    
    file_contents = None
    with open(filename, 'r') as f:
        file_contents = f.readlines()
    
    pointer = 0
    tape = [0]
    program = []
    curr_loc = 0
    count_brackets = 0
    jump_locs_rev = {}
    jump_locs_fwd = {}

    for line in file_contents:
        for char in line:
            if char == '\n':
                continue

            if char == '[':
                count_brackets += 1
            if char == ']':
                count_brackets -= 1

            program.append(char)

    if count_brackets != 0:
        print('Invalid bracket [] syntax')
        exit(0)
    else:
        loop_stack = []
        for i, char in enumerate(program, start=0):
            if char == '[':
                loop_stack.append(i)
            if char == ']':
                v = loop_stack.pop()
                jump_locs_rev[i] = v
                jump_locs_fwd[v] = i

    while curr_loc < len(program):
        char = program[curr_loc]
        if char == '>':
            pointer += 1
        if char == '<':
            pointer -= 1
        if char == '+':
            increment(tape, pointer)
        if char == '-':
            decrement(tape, pointer)
        if char == '.':
            print(chr(tape[pointer]), end='')
        if char == ',':
            tape[pointer] = int(input())
        
        if curr_loc in jump_locs_rev.keys():
            if not is_zero(tape, pointer):
                curr_loc = jump_locs_rev[curr_loc] - 1
        elif curr_loc in jump_locs_fwd.keys():
            if is_zero(tape, pointer):
                curr_loc = jump_locs_fwd[curr_loc]
        
        curr_loc += 1

if __name__ == '__main__':
    main()

File: js/bf.js
const fs = require("fs");
const readline = require("readline");

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
})

function change(tape, pointer, amount) {
    while (tape.length <= pointer) {
        tape.push(0);
    }

    tape[pointer] = (tape[pointer] + amount) % 255;
}

function increment(tape, pointer) {
    change(tape, pointer, 1);
}

function decrement(tape, pointer) {
    change(tape, pointer, -1);
}

function isZero(tape, pointer) {
    change(tape, pointer, 0);
    return tape[pointer] == 0;
}

function main() {
    if (process.argv.length != 3) {
        console.log("A simple BF interpreter.");
        console.log("Usage: \t node bf.js program.bf");
        process.exit(0);
    }

    const filename = process.argv[2];
    if (!fs.existsSync(filename)) {
        console.log(`Could not find file ${filename}`);
        process.exit(0);
    }

    const fileContents = fs.readFileSync(filename, "utf8");

    let pointer = 0;
    let tape = [0];
    let program = [];
    let currLoc = 0;
    let countBrackets = 0;
    let jumpLocsRev = {};
    let jumpLocsFwd = {};

    for (let line of fileContents) {
        for (let char of line) {
            if (char === "\n") {
                continue;
            }

            if (char === "[") {
                countBrackets++;
            }
            if (char === "]") {
                countBrackets--;
            }

            program.push(char);
        }
    }

    if (countBrackets !== 0) {
        console.log("Invalid bracket [] syntax");
        process.exit(0);
    }

    let loopStack = [];
    program.map((char, i) => {
        if (char === "[") {
            loopStack.push(i);
        }
        if (char === "]") {
            const v = loopStack.pop();
            jumpLocsRev[i] = v;
            jumpLocsFwd[v] = i;
        }
    });

    while (currLoc <= program.length) {
        const char = program[currLoc];
        if (char === ">") {
            pointer++;
        }
        if (char === "<") {
            pointer--;
        }
        if (char === "+") {
            increment(tape, pointer);
        }
        if (char === "-") {
            decrement(tape, pointer);
        }
        if (char === ".") {
            process.stdout.write(String.fromCharCode(tape[pointer]));
        }
        if (char === ",") {
            rl.question("", (input) => {
                tape[pointer] = Number(input);
            });
        }

        if (currLoc in jumpLocsRev) {
            if (!isZero(tape, pointer)) {
                currLoc = jumpLocsRev[currLoc] - 1;
            }
        } else if (currLoc in jumpLocsFwd) {
            if (isZero(tape, pointer)) {
                currLoc = jumpLocsFwd[currLoc];
            }
        }

        currLoc++;
    }
}

main();
process.exit();